source/extensions/filters/http/cache/cache_insert_queue.cc

```diff
@@ -69,22 +69,24 @@ class CacheInsertFragmentTrailers : public CacheInsertFragment {
 
 CacheInsertQueue::CacheInsertQueue(std::shared_ptr<HttpCache> cache,
                                    Http::StreamEncoderFilterCallbacks& encoder_callbacks,
-                                   InsertContextPtr insert_context, AbortInsertCallback abort)
+                                   InsertContextPtr insert_context, InsertQueueCallbacks& callbacks)
     : dispatcher_(encoder_callbacks.dispatcher()), insert_context_(std::move(insert_context)),
       low_watermark_bytes_(encoder_callbacks.encoderBufferLimit() / 2),
-      high_watermark_bytes_(encoder_callbacks.encoderBufferLimit()),
-      encoder_callbacks_(encoder_callbacks), abort_callback_(std::move(abort)), cache_(cache) {}
+      high_watermark_bytes_(encoder_callbacks.encoderBufferLimit()), callbacks_(callbacks),
+      cache_(cache) {}
 
 void CacheInsertQueue::insertHeaders(const Http::ResponseHeaderMap& response_headers,
                                      const ResponseMetadata& metadata, bool end_stream) {
   end_stream_queued_ = end_stream;
   // While zero isn't technically true for the size of headers, headers are
   // typically excluded from the stream buffer limit.
+  fragment_in_flight_ = true;
   insert_context_->insertHeaders(
       response_headers, metadata,
       [this, end_stream](bool cache_success) { onFragmentComplete(cache_success, end_stream, 0); },
       end_stream);
-  fragment_in_flight_ = true;
+  ASSERT(fragment_in_flight_,
+         "insertHeaders must post the callback to dispatcher, not just call it");
```

Comment:
The assumption is that *most* cache implementations will need to do something asynchronous, which means they'll be posting. It doesn't *specifically* care, but it would require implementing distinct code paths and more, different guards and things to ensure API correctness if direct callbacks were accepted.

This is the same problem many of the filter functions have, like you can't call `continueDecoding()` in a callback that *might be* called directly during `decodeHeaders()`. A previous cache implementation explicitly posted from within the callback to avoid that case; now, to avoid double-posting in the common case of a cache doing something async, the filter requires the cache to post, which is a performance penalty for the case where the cache is synchronous as in simple_http_cache, but should be an improvement in the common case.

(It's possible that the post requirement can go away because we're not doing continueDecoding any more, but I'd rather not change the API requirement until the rest of this is resolved and I'm sure it won't be important again.)

Reply:
this is a great PR comment. Can you summarize this as a code comment also?

Reply:
Done. (Though the summary is almost as long.)